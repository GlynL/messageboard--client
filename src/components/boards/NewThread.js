import React, { Component } from "react";
import { connect } from "react-redux";
import requireAuth from "../hocs/requireAuth";
import * as actions from "../../actions/threads";

const INITIAL_STATE = {
  title: "",
  text: ""
};

export class NewThread extends Component {
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  handleSubmit = e => {
    e.preventDefault();
    const board = this.props.match.params.board;
    // api uses board name not id
    this.props.addThread(this.state, board, this.props.token, () =>
      this.props.history.push(`/${board}`)
    );
  };

  handleChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  render() {
    return (
      <div>
        <p>{this.props.error}</p>
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="title">Title</label>
          <input
            onChange={this.handleChange}
            value={this.state.title}
            type="text"
            name="title"
            id="title"
          />
          <label htmlFor="text">Text</label>
          <textarea
            onChange={this.handleChange}
            value={this.state.text}
            name="text"
            id="text"
          />
          <button>Create Thread</button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: state.auth.authenticated,
  error: state.threads.errorMessage
});

export default connect(
  mapStateToProps,
  actions
)(requireAuth(NewThread));
