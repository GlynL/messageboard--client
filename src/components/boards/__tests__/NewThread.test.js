import React from "react";
import { shallow } from "enzyme";
import { NewThread } from "../NewThread";

describe("snapshots", () => {
  it("renders w/out error correctly", () => {
    const wrapper = shallow(<NewThread />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders w error correctly", () => {
    const wrapper = shallow(<NewThread error="test error" />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("input/textarea & form", () => {
  let wrapper, addThread;
  beforeEach(() => {
    addThread = jest.fn();
    wrapper = shallow(
      <NewThread
        addThread={addThread}
        match={{ params: { board: "test board" } }}
      />
    );
    wrapper.find("input").simulate("change", {
      target: { name: "title", value: "test input" }
    });
    wrapper
      .find("textarea")
      .simulate("change", { target: { name: "text", value: "test text" } });
    wrapper.update();
  });

  it("handles change of input", () => {
    expect(wrapper.find("input").prop("value")).toBe("test input");
  });

  it("handles change of textarea", () => {
    expect(wrapper.find("textarea").prop("value")).toBe("test text");
  });

  it("handles form submit", () => {
    const preventDefault = jest.fn();
    wrapper.find("form").simulate("submit", { preventDefault });
    wrapper.update();
    expect(preventDefault).toHaveBeenCalled();
    expect(addThread).toHaveBeenCalled();
  });
});
