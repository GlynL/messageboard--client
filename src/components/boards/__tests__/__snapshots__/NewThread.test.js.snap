// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders w error correctly 1`] = `
<div>
  <p>
    test error
  </p>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      id="title"
      name="title"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      htmlFor="text"
    >
      Text
    </label>
    <textarea
      id="text"
      name="text"
      onChange={[Function]}
      value=""
    />
    <button>
      Create Thread
    </button>
  </form>
</div>
`;

exports[`snapshots renders w/out error correctly 1`] = `
<div>
  <p />
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      id="title"
      name="title"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      htmlFor="text"
    >
      Text
    </label>
    <textarea
      id="text"
      name="text"
      onChange={[Function]}
      value=""
    />
    <button>
      Create Thread
    </button>
  </form>
</div>
`;
